add typeutil

svenwltr opened the pull request:
Adds a set data structure and helpers for pointers. Copied over from boombot.

svenwltr committed:
add dsutil

bjoernhaeuser commented on pkg/dsutil/pointer.go:
```diff
@@ -0,0 +1,22 @@
+
+// Pointer creates a pointer from the given value. This is useful, when
+// creating a pointer from a loop variable [1] or when creating a pointer from
+// a return value of another function, since both would require assing the
```
```suggestion
// a return value of another function, since both would require assigning the
```

bjoernhaeuser commented on pkg/dsutil/set.go:
```diff
@@ -0,0 +1,115 @@
+
+// Union adds each value from the given set to the set.
+func (s *Set[T]) Union(other Set[T]) {
+	for _, o := range other.ToList() {
```
No way to add directly without creating the list firs?

svenwltr replied:
Good point. Fixed in aebc957f13af60fe7ad86a2864b10a43ad6a1726

bjoernhaeuser reviewed (approved):
I would read `ds` as data science first, but thats just me, nice code

svenwltr commented:
I had to change the API a bit, because I think it is confusing to do union and intersection operations on an existing set. Therefore I created two functions which take multiple sets and return a new one.

svenwltr committed:
add slices to dsutil

svenwltr committed:
rename dsutil to typeutil

svenwltr commented:
I forgot what was blocking me in this PR. I tested it in Boombot and added some extra missing functions.

Also I renamed it to `typeutil` to not confuse it with `dsutil`.
